File: Ratelimiter/serverRequestThrottler.js
const fn3 = async () => {
  return new Promise((resolve) => {
    console.log("Starting task");
    setTimeout(
      async () => {
        console.log("Completed task");
        resolve();
      },
      Math.random() * 1000 * 2,
    );
  });
};
const fnInput = [fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3, fn3];
// Starting task
// Starting task
// Starting task
// Completed task
// Starting task
// Completed task
// Completed task
// Completed task





const limitConcurrency = async (tasks, concurrency = 3) => {
  let index = 0;

  const worker = async () => {
    while (index < tasks.length) {
      const currentTask = tasks[index++];
      await currentTask();  // Execute and wait
    }
  };

  // Create "concurrency" number of workers
  const workers = Array.from({ length: concurrency }, worker);
console.log(workers);
  // Wait for all workers to finish
  await Promise.all(workers);
};

limitConcurrency(fnInput, 3);

File: Go/Ratelimiter-go/server_request_throttler.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

func fn3() {
	fmt.Println("Starting task")
	delay := time.Duration(rand.Float64()*2000) * time.Millisecond
	time.Sleep(delay)
	fmt.Println("Completed task")
}

func limitConcurrency(tasks []func(), concurrency int) {
	var index int
	var mu sync.Mutex

	worker := func(wg *sync.WaitGroup) {
		defer wg.Done()
		for {
			mu.Lock()
			if index >= len(tasks) {
				mu.Unlock()
				return
			}
			current := index
			index++
			mu.Unlock()
			tasks[current]()
		}
	}

	var wg sync.WaitGroup
	for i := 0; i < concurrency; i++ {
		wg.Add(1)
		go worker(&wg)
	}
	wg.Wait()
}

func demoConcurrencyThrottler() {
	tasks := make([]func(), 24)
	for i := range tasks {
		tasks[i] = fn3
	}
	limitConcurrency(tasks, 3)
}
